### PR TITLE
ColorPicker: adds support for keyboard navigation

### LINKED ISSUE #42525: ColorPickerPopover is non-navigable using keyboard
it's very difficult for users to navigate or select colors using the keyboard. 

**Steps to reproduce**:
Start by creating a timeSeries panel and navigating to the panelEdit. In panelEdit, navigate to the thresholds section. You'll find a color picker (circled in red). Click it and try navigating the colorSwatches using tab. Try clicking a color as well using the keyboard.
**Actual Result**:
ColorPickerPopover is not navigable using keyboard
**Expected Result**
ColorPickerPopover is navigable using keyboard
**Relevant WCAG Criteria:** [#.#.# WCAG Criterion](link to https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0)

**Environment**: 
- Grafana version: Grafana v8.4.0-pre (f454a5ce60)
- Data source type & version:
- User OS & Browser: MacOS, chrome
- Others:

### LINKED ISSUE #42526: PanelEdit: ColorPickerPopover cannot be opened when Enter key is pressed.
**Steps to reproduce**:
Create a timeSeries panel and navigate to the panelEdit. In panelEdit, navigate to the thresholds section. You'll find a color picker (circled in red). Try opening it by pressing `Enter` key. 
**Actual Result**:
ColorPickerPopover is not opened.
**Expected Result**
ColorPickerPopover is opened.
**Relevant WCAG Criteria:** [#.#.# WCAG Criterion](link to https://www.w3.org/WAI/WCAG21/quickref/?versions=2.0)

**Environment**:
- Grafana version: Grafana v8.4.0-pre (f454a5ce60)
- Data source type & version:
- User OS & Browser: MacOs, Chrome
- Others:

### PR DESCRIPTION
**Pull request by thisisobate**
**What this PR does / why we need it**:
This PR adds support for keyboard navigation in ColorPicker widget.
**Which issue(s) this PR fixes**:

Fixes #42525 
Fixes #42526

**Special notes for your reviewer**:

### CODE REVIEW

File: packages/grafana-ui/src/components/ColorPicker/NamedColorsGroup.tsx [hunk 1]
```diff
@@ -46,6 +47,7 @@ const NamedColorsGroup: FunctionComponent<NamedColorsGroupProps> = ({
                   isSelected={shade.name === selectedColor}
                   color={shade.color}
                   onClick={() => onColorSelect(shade.name)}
+                  onKeyDown={() => onColorSelect(primaryShade.name)}
```
Comment (hugohaggmark) [inline packages/grafana-ui/src/components/ColorPicker/NamedColorsGroup.tsx:50]: Should this be
```suggestion
                  onKeyDown={() => onColorSelect(shade.name)}
```

**Review by hugohaggmark — COMMENTED** [review]
Good work, I have some thoughts. With the current implementation any keydown event will cause the NamedColorsGroup to change color, not sure that is something we want.

Also, what are your thoughts on the user experience? If I want a Purple:ish color I've to tab a lot of times to get there.

File: packages/grafana-ui/src/components/ColorPicker/NamedColorsGroup.tsx (hunk 1 above)
Comment (hugohaggmark) [reply packages/grafana-ui/src/components/ColorPicker/NamedColorsGroup.tsx:50]: I think the current implementation has a bug because of the above 😉

Comment (thisisobate) [reply packages/grafana-ui/src/components/ColorPicker/NamedColorsGroup.tsx:50]: > With the current implementation any keydown event will cause the NamedColorsGroup to change color, not sure that is something we want.

True...would push a fix now

Comment (thisisobate) [reply packages/grafana-ui/src/components/ColorPicker/NamedColorsGroup.tsx:50]: > Also, what are your thoughts on the user experience? If I want a Purple:ish color I've to tab a lot of times to get there.

Yeah...I was actually considering implementing just tabs within this current scope of work. Definitely there's some work to be done UX-wise (e.g adding arrow navigation) but I'm considering doing that in a separate PR.

**Comment by thisisobate** [comment]
> With the current implementation any keydown event will cause the NamedColorsGroup to change color, not sure that is something we want.

True...would push a fix now

**Comment by thisisobate** [comment]
> Also, what are your thoughts on the user experience? If I want a Purple:ish color I've to tab a lot of times to get there.

Yeah...I was actually considering implementing just tabs within this current scope of work. Definitely there's some work to be done UX-wise (e.g adding arrow navigation) but I'm considering doing that in a separate PR.

File: packages/grafana-ui/src/components/ColorPicker/NamedColorsGroup.tsx (hunk 1 above)
Comment (thisisobate) [reply packages/grafana-ui/src/components/ColorPicker/NamedColorsGroup.tsx]: Great catch! Thanks!

Comment (hugohaggmark) [reply packages/grafana-ui/src/components/ColorPicker/NamedColorsGroup.tsx]: Are you aware that you can commit suggestions directly in the Github UI?

**Comment by hugohaggmark** [comment]
> > Also, what are your thoughts on the user experience? If I want a Purple:ish color I've to tab a lot of times to get there.
> 
> Yeah...I was actually considering implementing just tabs within this current scope of work. Definitely there's some work to be done UX-wise (e.g adding arrow navigation) but I'm considering doing that in a separate PR.

Hmm, I don't understand this thing with pushing things to the next PR. This is the PR that adds support for keyboard navigation, the title says so. I think you should consider what the keyboard navigation should be for this component as part of this PR not in a follow up PR.

**Review by ashharrison90 — COMMENTED** [review]
maybe i'm missing something, but is there any reason we can't use `react-aria` here? `useOverlay`/`useOverlayTrigger` should handle all this behaviour, and we can probably use `useTabs` for the 2 tabs (although that can probably wait since we've never used it before)?

**Comment by thisisobate** [comment]
@ashharrison90 I think we're majorly concerned about the user flow when navigating using either tabs or arrow keys. Considering bringing this up in the next UX feedback session.

File: packages/grafana-ui/src/components/ColorPicker/ColorPicker.tsx [hunk 2]
```diff
@@ -38,6 +38,27 @@ export const colorPickerFactory = <T extends ColorPickerProps>(
...
+
+      return;
+    };
+
+    activatePopper = (event: React.KeyboardEvent<HTMLDivElement>, showPopper: () => void) => {
+      if (event.key === 'Enter') {
+        showPopper();
+      }
+      return;
+    };
```
Comment (ashharrison90) [inline packages/grafana-ui/src/components/ColorPicker/ColorPicker.tsx:60]: @thisisobate sorry, don't quite understand. i'm trying to say that i don't think we need all this logic as `react-aria` will handle this in `useOverlay`/`useOverlayTrigger`. does that make sense?

Comment (thisisobate) [reply packages/grafana-ui/src/components/ColorPicker/ColorPicker.tsx:60]: yeah...it does.

**Comment by kaydelaney** [comment]
Nice work on this! Some thoughts:
- Typically for [composite widgets](https://w3c.github.io/aria/#composite), the rule-of-thumb is for the arrow keys to navigate within the widget, rather than tab/shift+tab
- The way pressing tab selects the previously focused color is maybe a little confusing. It might be better to have the currently focused color be the selected one.
- It doesn't seem like "transparent" or "text color" can be selected through keyboard navigation.
- Clicking the button to bring up the color picker for the single color color scheme seems to immediately get rid of focus:

https://user-images.githubusercontent.com/45561153/144580316-3e2fc49a-db6c-4ac4-b177-fd4363bc155a.mp4

Let me know if you need me to clarify any of this!

**Comment by thisisobate** [comment]
Thanks for the feedback @kaydelaney 

Regarding this particular component, we decided to create a new design that would aid keyboard accessibility. Would create a new PR once the implementation is ready.

Would proceed to close this one in favor of the new one.

**Pull request closed without merging**